src/osp/Active/SysNewton.cpp

```diff
@@ -9,35 +9,47 @@ using namespace osp::active;
 
 void cb_force_torque(const NewtonBody* body, dFloat timestep, int threadIndex)
 {
-    SysNewton* sysNewton = static_cast<SysNewton*>(
+    auto* sysNewton = static_cast<SysNewton*>(
                 NewtonWorldGetUserData(NewtonBodyGetWorld(body)));
     ActiveScene &scene = sysNewton->get_scene();
 
-    ACompNwtBody *bodyComp
-            = static_cast<ACompNwtBody*>(NewtonBodyGetUserData(body));
+    auto *bodyComp = static_cast<ACompNwtBody*>(NewtonBodyGetUserData(body));
     DataPhyRigidBody &bodyPhy = bodyComp->m_bodyData;
 
-    //Matrix4 matrix;
-
-    ACompTransform& transform = scene.get_registry()
-                                        .get<ACompTransform>(bodyComp->m_entity);
-    NewtonBodyGetMatrix(body, transform.m_transform.data());
+    auto& transform
+            = scene.get_registry().get<ACompTransform>(bodyComp->m_entity);
 
+    // Get new transform matrix from newton
     Matrix4 matrix;
     NewtonBodyGetMatrix(body, matrix.data());
 
-    // Check if floating origin translation is in progress
-    if (scene.floating_origin_in_progress())
+    // Check if transform is being set
+    if (transform.m_transformDirty)
     {
-        // Do floating origin translation, then set position
+        // Instead of setting transform directly, it's better to add the
+        // difference with the previous transform. This ensures that the changes
+        // to the rigid body from Newton this frame will still apply.
+
+        // TODO: there's probably a better way to do this. if not, account for
+        //       rotation too, since this is translation only.
 
-        matrix[3].xyz() += scene.floating_origin_get_total();
+        Vector3 diffTranslate = transform.m_transform.translation()
+                              - bodyComp->m_prevTransform.translation();
+        matrix.translation() += diffTranslate;
```

Comment:
This callback runs before the actual position is updated, correct? Otherwise this would cause a frame of motion to be dropped.

Reply:
Yes, but this was janky and removed at the latest commit. The positions are now updated after the physics update in the update_world() function in a separate loop.